apps/nsqd/nsqd.go

```diff
@@ -218,31 +221,41 @@ func (p *program) Start() error {
 	if configFile != "" {
 		_, err := toml.DecodeFile(configFile, &cfg)
 		if err != nil {
-			log.Fatalf("ERROR: failed to load config file %s - %s", configFile, err.Error())
+			lg.LogFatal("[nsqd] ", "failed to load config file %s - %s", configFile, err)
 		}
 	}
 	cfg.Validate()
 
 	options.Resolve(opts, flagSet, cfg)
-	nsqd := nsqd.New(opts)
+	nsqd, err := nsqd.New(opts)
+	if err != nil {
+		lg.LogFatal("[nsqd] ", "failed to create nsqd instance - %s", err)
+	}
+	p.nsqd = nsqd
 
-	err := nsqd.LoadMetadata()
+	err = p.nsqd.LoadMetadata()
 	if err != nil {
-		log.Fatalf("ERROR: %s", err.Error())
+		lg.LogFatal("[nsqd] ", "failed to load metadata - %s", err)
 	}
-	err = nsqd.PersistMetadata()
+	err = p.nsqd.PersistMetadata()
 	if err != nil {
-		log.Fatalf("ERROR: failed to persist metadata - %s", err.Error())
+		lg.LogFatal("[nsqd] ", "failed to persist metadata - %s", err)
 	}
-	nsqd.Main()
 
-	p.nsqd = nsqd
+	go func() {
+		err := p.nsqd.Main()
+		if err != nil {
+			p.Stop()
+			lg.LogFatal("[nsqd] ", "%s", err)
```

Comment:
might want to log before `Stop()` so error is logged closer to when it happened
(could log in the `once.Do(func() {exitCh <- err})` and then again here at fatal exit)

Reply:
Yea, was just being lazy because `lg.LogFatal` handles the `os.Exit(1)` as implemented... For the record, I'm unhappy with how these logging changes ended up — they're an improvement from what was there, but also ugly in some ways, so I'm open to ideas if you have any.